libs/vgc/tools/sculpt.cpp

```diff
@@ -103,11 +103,12 @@ class SculptGrabAction : public ui::Action {
             return;
         }
 
-        canvas::Canvas* canvas = tool_->canvas();
-        workspace::Workspace* workspace = tool_->workspace();
-        if (!canvas || !workspace) {
+        auto context = tool_->contextLock();
```

Comment:
Missing check for dereferencing `tool_` here. In fact, `tools_` is currently defined as a raw pointer, should be converted to WeakPtr too.